tests/SecurityGroup.py

```diff
@@ -0,0 +1,127 @@
+        self.publishTrans = None
+
+    @staticmethod
+    def __lessThan(lhsNode, rhsNode):
+        assert lhsNode != rhsNode
+        return lhsNode.getParticipant() < rhsNode.getParticipant()
+
+    def __consistentStorage(self):
+        intersect = list(set([x.getParticipant() for x in self.participants]) & set([x.getParticipant() for x in self.nonParticipants]))
+        if len(intersect) != 0:
+            Utils.exitError("Values: {}, are in the nonParticipants list: {}, and the participants list: {}".format(", ".join(intersect), SecurityGroup.createAction(self.nonParticipants), SecurityGroup.createAction(self.participants)))
+
+    def __addParticipants(self, nodes):
+        if Utils.Debug: Utils.Print("Moving the following: {}, from the nonParticipants list: {}, to the participants list: {}".format(SecurityGroup.createAction(nodes), SecurityGroup.createAction(self.nonParticipants), SecurityGroup.createAction(self.participants)))
+        for node in nodes:
+            if node not in self.nonParticipants:
+                for nonParticipant in self.nonParticipants:
+                    match = "" if node == nonParticipant else " not"
+                    Utils.Print("node: '{}' does{} match '{}'".format(node, match, nonParticipant))
+            assert node in self.nonParticipants, "Cannot remove {} from nonParticipants list: {}".format(node, SecurityGroup.createAction(self.nonParticipants))
+            self.nonParticipants.remove(node)
+
+        self.participants.extend(nodes)
+        if Utils.Debug: Utils.Print("nonParticipants list: {}, to the participants list: {}".format(SecurityGroup.createAction(self.nonParticipants), SecurityGroup.createAction(self.participants)))
+        self.__consistentStorage()
+
+    def __remParticipants(self, nodes):
+        if Utils.Debug: Utils.Print("Moving the following: {}, from the participants list: {}, to the non-participants list: {}".format(SecurityGroup.createAction(nodes), SecurityGroup.createAction(self.participants), SecurityGroup.createAction(self.nonParticipants)))
+        for node in nodes:
+            self.participants.remove(node)
+
+        self.nonParticipants.extend(nodes)
+        if Utils.Debug: Utils.Print("participants list: {}, to the non-participants list: {}".format(SecurityGroup.createAction(self.participants), SecurityGroup.createAction(self.nonParticipants)))
+        self.__consistentStorage()
+
+    @staticmethod
+    def createAction(nodes):
+        return None if len(nodes) == 0 else \
+            "[[{}]]".format(','.join(['"{}"'.format(node.getParticipant()) for node in nodes]))
+
+    def editSecurityGroup(self, addNodes=[], removeNodes=[]):
+
+        addAction = SecurityGroup.createAction(addNodes)
+        # doing deep copy in case the passed in list IS participants or nonParticipants lists, which will be adjusted
+        self.__addParticipants(copy.deepcopy(addNodes))
```

Comment:
I don't understand why do we use `deepcopy` here? So it will copy `Node` instance. Why do we need this?
passed list is not stored anywhere inside `__addParticipants`, its elements are added to class members.

Reply:
so that calls like securityGroup.editSecurityGroup(securityGroup.nonParticipating) don't change addNodes when it calls remove self.nonParticipating, which otherwise would be the exact same container.